File: robert-server-ws-rest/src/test/java/fr/gouv/stopc/robertserver/ws/test/GrpcMockManager.java
package fr.gouv.stopc.robertserver.ws.test;

import com.google.protobuf.ByteString;
import fr.gouv.stopc.robert.crypto.grpc.server.messaging.*;
import io.grpc.Server;
import io.grpc.ServerBuilder;
import io.grpc.stub.StreamObserver;
import lombok.SneakyThrows;
import org.mockito.Mockito;
import org.springframework.lang.NonNull;
import org.springframework.test.context.TestContext;
import org.springframework.test.context.TestExecutionListener;

import java.io.IOException;
import java.util.*;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doThrow;

public class GrpcMockManager implements TestExecutionListener {

    private static final CryptoGrpcStub CRYPTO_GRPC_STUB = Mockito.spy(new CryptoGrpcStub());

    private static final Server GRPC_MOCK = ServerBuilder.forPort(0)
            .addService(CRYPTO_GRPC_STUB)
            .build();

    static {
        try {
            GRPC_MOCK.start();
        } catch (IOException e) {
            throw new RuntimeException("Unable to start GRPC mock", e);
        }
        System.setProperty("robert.crypto.server.host", "localhost");
        System.setProperty("robert.crypto.server.port", String.valueOf(GRPC_MOCK.getPort()));
    }

    @Override
    @SneakyThrows
    public void beforeTestMethod(@NonNull TestContext testContext) {
        Mockito.reset(CRYPTO_GRPC_STUB);
        CryptoGrpcStub.reset();
    }

    @SneakyThrows
    public static void givenCryptoServerIsOffline() {
        final var error = new RuntimeException("Crypto server is offline");
        doThrow(error).when(CRYPTO_GRPC_STUB).createRegistration(any(), any());
        doThrow(error).when(CRYPTO_GRPC_STUB).getIdFromAuth(any(), any());
        doThrow(error).when(CRYPTO_GRPC_STUB).getIdFromStatus(any(), any());
        doThrow(error).when(CRYPTO_GRPC_STUB).deleteId(any(), any());
    }

    public static void givenCryptoServerRaiseMissingIdentityForEbid(String ebid) {
        CryptoGrpcStub.RESPONSE_STATUS_BY_EBID.put(ebid, 404);
    }

    public static void givenCryptoServerRaiseMissingDailyKeyForEbid(String ebid) {
        CryptoGrpcStub.RESPONSE_STATUS_BY_EBID.put(ebid, 430);
    }

    public static void givenCryptoServerRaiseError400ForEbid(String ebid) {
        CryptoGrpcStub.RESPONSE_STATUS_BY_EBID.put(ebid, 400);
    }

    public static void verifyNoInteractionsWithCryptoServer() {
        Mockito.verifyNoInteractions(CRYPTO_GRPC_STUB);
    }

    private static class CryptoGrpcStub extends CryptoGrpcServiceImplGrpc.CryptoGrpcServiceImplImplBase {

        private static final Map<String, Integer> RESPONSE_STATUS_BY_EBID = new HashMap<>();

        private static void reset() {
            RESPONSE_STATUS_BY_EBID.clear();
        }

        private Optional<ErrorMessage> handleConfiguredError(ByteString requestEbid, ByteString requestMac) {
            if (RESPONSE_STATUS_BY_EBID.containsKey(requestEbid.toStringUtf8())) {
                return Optional.of(
                        ErrorMessage.newBuilder()
                                .setCode(RESPONSE_STATUS_BY_EBID.get(requestEbid.toStringUtf8()))
                                .setDescription("Some error description")
                                .build()
                );
            } else {
                return Optional.empty();
            }
        }

        @Override
        public void createRegistration(CreateRegistrationRequest request,
                StreamObserver<CreateRegistrationResponse> responseObserver) {
            final var clientPublicKey = request.getClientPublicKey().toStringUtf8();
            responseObserver.onNext(
                    CreateRegistrationResponse.newBuilder()
                            .setIdA(ByteString.copyFromUtf8("fake idA for " + clientPublicKey))
                            .setTuples(ByteString.copyFromUtf8("fake encrypted tuples for " + clientPublicKey))
                            .build()
            );
            responseObserver.onCompleted();
        }

        @Override
        public void getIdFromAuth(GetIdFromAuthRequest request,
                StreamObserver<GetIdFromAuthResponse> responseObserver) {
            final var response = handleConfiguredError(request.getEbid(), request.getMac())
                    .map(
                            err -> GetIdFromAuthResponse.newBuilder()
                                    .setError(err)
                                    .build()
                    )
                    .orElse(
                            GetIdFromAuthResponse.newBuilder()
                                    .setEpochId(request.getEpochId())
                                    .setIdA(request.getEbid())
                                    .build()
                    );
            responseObserver.onNext(response);
            responseObserver.onCompleted();
        }

        @Override
        public void getIdFromStatus(GetIdFromStatusRequest request,
                StreamObserver<GetIdFromStatusResponse> responseObserver) {
            final var response = handleConfiguredError(request.getEbid(), request.getMac())
                    .map(
                            err -> GetIdFromStatusResponse.newBuilder()
                                    .setError(err)
                                    .build()
                    )
                    .orElse(
                            GetIdFromStatusResponse.newBuilder()
                                    .setEpochId(request.getEpochId())
                                    .setIdA(request.getEbid())
                                    .setTuples(
                                            ByteString.copyFromUtf8(
                                                    "fake encrypted tuples for " + request.getEbid().toStringUtf8()
                                            )
                                    )
                                    .build()
                    );
            responseObserver.onNext(response);
            responseObserver.onCompleted();
        }

        @Override
        public void deleteId(DeleteIdRequest request, StreamObserver<DeleteIdResponse> responseObserver) {
            final var response = handleConfiguredError(request.getEbid(), request.getMac())
                    .map(
                            err -> DeleteIdResponse.newBuilder()
                                    .setError(err)
                                    .build()
                    )
                    .orElse(
                            DeleteIdResponse.newBuilder()
                                    .setIdA(request.getEbid())
                                    .build()
                    );
            responseObserver.onNext(response);
            responseObserver.onCompleted();
        }
    }
}

File: robert-server-ws-rest/src/test/kotlin/fr/gouv/stopc/robertserver/ws/test/GrpcMockManager.kt
package fr.gouv.stopc.robertserver.ws.test

import com.google.protobuf.ByteString
import fr.gouv.stopc.robert.crypto.grpc.server.messaging.CreateRegistrationRequest
import fr.gouv.stopc.robert.crypto.grpc.server.messaging.CreateRegistrationResponse
import fr.gouv.stopc.robert.crypto.grpc.server.messaging.CryptoGrpcServiceImplGrpc.CryptoGrpcServiceImplImplBase
import fr.gouv.stopc.robert.crypto.grpc.server.messaging.DeleteIdRequest
import fr.gouv.stopc.robert.crypto.grpc.server.messaging.DeleteIdResponse
import fr.gouv.stopc.robert.crypto.grpc.server.messaging.ErrorMessage
import fr.gouv.stopc.robert.crypto.grpc.server.messaging.GetIdFromAuthRequest
import fr.gouv.stopc.robert.crypto.grpc.server.messaging.GetIdFromAuthResponse
import fr.gouv.stopc.robert.crypto.grpc.server.messaging.GetIdFromStatusRequest
import fr.gouv.stopc.robert.crypto.grpc.server.messaging.GetIdFromStatusResponse
import io.grpc.ServerBuilder
import io.grpc.stub.StreamObserver
import lombok.SneakyThrows
import org.mockito.Mockito
import org.mockito.Mockito.doThrow
import org.mockito.kotlin.anyOrNull
import org.springframework.lang.NonNull
import org.springframework.test.context.TestContext
import org.springframework.test.context.TestExecutionListener

class GrpcMockManager : TestExecutionListener {
    @SneakyThrows
    override fun beforeTestMethod(@NonNull testContext: TestContext) {
        Mockito.reset(CRYPTO_GRPC_STUB)
        CryptoGrpcStub.reset()
    }

    companion object {
        private val CRYPTO_GRPC_STUB = Mockito.spy(CryptoGrpcStub())
        private val GRPC_MOCK = ServerBuilder.forPort(0)
            .addService(CRYPTO_GRPC_STUB)
            .build()
            .apply {
                start()
                System.setProperty("robert-ws.crypto-server-uri", "dns:///localhost:$port")
            }

        fun givenCryptoServerIsOffline() {
            val error = RuntimeException("Crypto server is offline")
            doThrow(error).`when`(CRYPTO_GRPC_STUB).createRegistration(anyOrNull(), anyOrNull())
            doThrow(error).`when`(CRYPTO_GRPC_STUB).getIdFromAuth(anyOrNull(), anyOrNull())
            doThrow(error).`when`(CRYPTO_GRPC_STUB).getIdFromStatus(anyOrNull(), anyOrNull())
            doThrow(error).`when`(CRYPTO_GRPC_STUB).deleteId(anyOrNull(), anyOrNull())
        }

        fun givenCryptoServerRaiseMissingIdentityForEbid(ebid: String) {
            CryptoGrpcStub.RESPONSE_STATUS_BY_EBID[ebid] = 404
        }

        fun givenCryptoServerRaiseMissingDailyKeyForEbid(ebid: String) {
            CryptoGrpcStub.RESPONSE_STATUS_BY_EBID[ebid] = 430
        }

        fun givenCryptoServerRaiseError400ForEbid(ebid: String) {
            CryptoGrpcStub.RESPONSE_STATUS_BY_EBID[ebid] = 400
        }

        fun verifyNoInteractionsWithCryptoServer() = Mockito.verifyNoInteractions(CRYPTO_GRPC_STUB)
    }

    private open class CryptoGrpcStub : CryptoGrpcServiceImplImplBase() {

        companion object {
            val RESPONSE_STATUS_BY_EBID: MutableMap<String, Int> = HashMap()
            fun reset() {
                RESPONSE_STATUS_BY_EBID.clear()
            }
        }

        private fun handleConfiguredError(requestEbid: ByteString, requestMac: ByteString): ErrorMessage? {
            return if (RESPONSE_STATUS_BY_EBID.containsKey(requestEbid.toStringUtf8())) {
                ErrorMessage.newBuilder()
                    .setCode(RESPONSE_STATUS_BY_EBID[requestEbid.toStringUtf8()]!!)
                    .setDescription("Some error description")
                    .build()
            } else {
                null
            }
        }

        override fun createRegistration(
            request: CreateRegistrationRequest,
            responseObserver: StreamObserver<CreateRegistrationResponse>
        ) {
            val idA = request.clientPublicKey.toStringUtf8().takeLast(5)
            responseObserver.onNext(
                CreateRegistrationResponse.newBuilder()
                    .setIdA(ByteString.copyFromUtf8(idA))
                    .setTuples(ByteString.copyFromUtf8("fake encrypted tuples for '$idA'"))
                    .build()
            )
            responseObserver.onCompleted()
        }

        override fun getIdFromAuth(
            request: GetIdFromAuthRequest,
            responseObserver: StreamObserver<GetIdFromAuthResponse>
        ) {
            val err = handleConfiguredError(request.ebid, request.mac)
            val response = if (null != err) {
                GetIdFromAuthResponse.newBuilder()
                    .setError(err)
                    .build()
            } else {
                GetIdFromAuthResponse.newBuilder()
                    .setEpochId(request.epochId)
                    .setIdA(request.ebid.take(5))
                    .build()
            }
            responseObserver.onNext(response)
            responseObserver.onCompleted()
        }

        override fun getIdFromStatus(
            request: GetIdFromStatusRequest,
            responseObserver: StreamObserver<GetIdFromStatusResponse>
        ) {
            val err = handleConfiguredError(request.ebid, request.mac)
            val response = if (null != err) {
                GetIdFromStatusResponse.newBuilder()
                    .setError(err)
                    .build()
            } else {
                val idA = request.ebid.take(5)
                GetIdFromStatusResponse.newBuilder()
                    .setEpochId(request.epochId)
                    .setIdA(idA)
                    .setTuples(
                        ByteString.copyFromUtf8("fake encrypted tuples for '${idA.toStringUtf8()}'")
                    )
                    .build()
            }
            responseObserver.onNext(response)
            responseObserver.onCompleted()
        }

        override fun deleteId(request: DeleteIdRequest, responseObserver: StreamObserver<DeleteIdResponse>) {
            val err = handleConfiguredError(request.ebid, request.mac)
            val response = if (null != err) {
                DeleteIdResponse.newBuilder()
                    .setError(err)
                    .build()
            } else {
                DeleteIdResponse.newBuilder()
                    .setIdA(request.ebid.take(5))
                    .build()
            }
            responseObserver.onNext(response)
            responseObserver.onCompleted()
        }

        private fun ByteString.take(n: Int) = ByteString.copyFrom(this.toByteArray().take(n).toByteArray())
    }
}
